### PR TITLE
Add update-endpoint

### LINKED ISSUE #233: Think if giving consent twice works as a rectification or not
Giving consent a second time allows the subject to update his personal data. If his data is updated, that may count as a rectification, and we may need to handle that also as a rectification.
Low priority

### LINKED ISSUE #253: Issue paging rectification requests
**Steps to reproduce:**
1. Add at least 10 rectification requests using `/api/subject/initiate-rectification`, example body:
```
{
	"rectificationPayload": {
		"firstName": "Maks",
		"email": "maksymilian.majer@somefakedomain.biz",
		"someKey": "Some data goes here"
	},
	"requestReason": "The old data was incorrect"
}
```
2. Make request to `/api/management/subjects/rectification-requests/list?page=1`
3. Make request to `/api/management/subjects/rectification-requests/list?page=2`

**Expected result:**
The second page should look like:
```
{
    "data": [
        {
            "id": 11,
            "request_reason": "The data was incorrect two.",
            "created_at": "2018-07-02T21:31:37.440Z"
        },
        {
            "id": 12,
            "request_reason": "The data was incorrect two three.",
            "created_at": "2018-07-02T21:31:43.530Z"
        },
        {
            "id": 13
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:50.452Z"
        },
    ],
    "paging": {
        "current": 2,
        "total": 2
    }
}
```

**Actual result:**
The second page is offset by only 1 item:
```
{
    "data": [
        {
            "id": 2,
            "request_reason": "The data was incorrect two.",
            "created_at": "2018-07-02T21:31:37.440Z"
        },
        {
            "id": 3,
            "request_reason": "The data was incorrect two three.",
            "created_at": "2018-07-02T21:31:43.530Z"
        },
        {
            "id": 4,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:50.452Z"
        },
        {
            "id": 5,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:51.824Z"
        },
        {
            "id": 6,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:52.611Z"
        },
        {
            "id": 7,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:53.295Z"
        },
        {
            "id": 8,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:53.947Z"
        },
        {
            "id": 9,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:54.586Z"
        },
        {
            "id": 10,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:55.189Z"
        },
        {
            "id": 11,
            "request_reason": "The data was incorrect two next",
            "created_at": "2018-07-04T14:02:55.793Z"
        }
    ],
    "paging": {
        "current": 2,
        "total": 2
    }
}
```

### LINKED ISSUE #257: Create a separate endpoint to update consented processors
We are allowing the subject to update his data through a give-consent request, but this should not happen since we have an endpoint, initiate-rectification, just for that. This enpoint should look like:

POST
/api/subjects/update-consent
body: processors[]
(only the new processors consented in the body)

It also requires changes in the give-consent endpoint to block a subject from updating his data by calling that endpoint more than once.

### PR DESCRIPTION
**Pull request by sindelio**
Closes #257 Closes #233 Closes #253 
Add update-consent endpoint, it's validator and some necessary changes in the give-consent endpoint.
Also fixes the pagination issues in the rectification endpoints.
Tested in Postman, and the test suite is updated and working.
It's ready to receive final reviews!

### CODE REVIEW

File: cg/src/domains/management/subjects/subjects.service.js [hunk 1]
```diff
@@ -81,6 +64,23 @@ class SubjectsService {
     };
   }
 
+  async getRequestData(id) {
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.service.js]: why did you move this?

File: cg/src/domains/subjects/subjects.service.js [hunk 2]
```diff
@@ -28,28 +32,51 @@ class SubjectsService {
     });
   }
 
-  async registerConsentToProcessData(subjectId, personalData, processorIds = []) {
+  async registerConsentToProcessData(subjectId, personalData, processorsConsented = []) {
+    const [ subjectExists ] = await this.db('subjects')
+      .where('id', subjectId);
+    
+    if(subjectExists) throw new Unauthorized('Subject already gave consent, please use the update-consent endpoint');
```
Comment (Jakedavies) [inline cg/src/domains/subjects/subjects.service.js]: This error should probably just be "Subject has already given consent"

File: cg/src/domains/subjects/subjects.service.js [hunk 3]
```diff
@@ -28,28 +32,51 @@ class SubjectsService {
...
 
-      await recordConsentGivenTo(subjectId, processorIdsWithAddresses.map(p => p.address));
+  async updateConsent(subjectId, processorsConsented = []){
+    const [ subjectExists ] = await this.db('subjects')
+      .where('id', subjectId);
+  
+    if(!subjectExists) throw new NotFound('Subject not found, please use the give-consent endpoint');
+    let processorIdsWithAddresses;
+    await this.db.transaction(async trx => {
+      processorIdsWithAddresses = await this._getProcessorIdsWithAddresses(trx, processorsConsented);
```
Comment (Jakedavies) [inline cg/src/domains/subjects/subjects.service.js]: can we share this validation logic with the above function? abstract it into a shared function call?

File: cg/src/domains/management/subjects/subjects.service.js (hunk 1 above)
Comment (sindelio) [reply cg/src/domains/management/subjects/subjects.service.js]: I'm sure there was a reason xD
I'm moving it close to where it is being used now

**Comment by sindelio** [comment]
After some fighting with rebasing, it's good to receive reviews o/

File: cg/src/domains/subjects/subjects.service.js [hunk 4]
```diff
@@ -27,21 +31,48 @@ class SubjectsService {
...
+
+  async updateConsent(subjectId, processorsConsented = []){
+    const [ subjectExists ] = await this.db('subjects')
+      .where('id', subjectId);
+    
+    if(!subjectExists) throw new NotFound('Subject not found');
+    let processorIdsWithAddresses;
+    await this.db.transaction(async trx => {
+      processorIdsWithAddresses = await this._getProcessorIdsWithAddresses(trx, processorsConsented);
+      if (processorIdsWithAddresses.length !== processorsConsented.length) {
```
Comment (maksymilian-majer) [inline cg/src/domains/subjects/subjects.service.js]: There still is this code duplication @Jakedavies mentioned

File: cg/test/management/subjects.api.spec.js [hunk 5]
```diff
@@ -591,7 +591,7 @@ describe('List rectification requests', () => {
   });
 });
 
-describe('Rectification requests archive', () => {
```
Comment (maksymilian-majer) [inline cg/test/management/subjects.api.spec.js]: This was named like that on purpose, so the tests read like:
Rectification requests archive + Should list archived requests in pages bigger than 1 correctly

Now they read like:
Tests of listing archived rectification requests + Should list archived requests in pages bigger than 1 correctly

This should be more like natural sentences. The tests should only test, but the endpoints/class methods/functions should do whatever we're testing.

File: cg/src/domains/management/subjects/subjects.service.js [hunk 6]
```diff
@@ -57,62 +57,54 @@ class SubjectsService {
...
-
     const requests = await this.db('rectification_requests')
       .select('id')
       .select('request_reason')
       .select('rectification_requests.created_at')
       .where('status', RECTIFICATION_STATUSES.PENDING)
       .join('subject_keys', 'rectification_requests.subject_id', 'subject_keys.subject_id')
       .limit(PAGE_SIZE)
-      .offset(page - 1);
+      .offset((requestedPage - 1 ) * PAGE_SIZE);
```
[the remark this answers is not in the document]
Comment (sindelio) [reply cg/src/domains/management/subjects/subjects.service.js]: Yeah my bad, forgot to make another branch

File: cg/src/domains/subjects/subjects.service.js (hunk 4 above)
Comment (sindelio) [reply cg/src/domains/subjects/subjects.service.js]: I will address it now

File: cg/test/management/subjects.api.spec.js (hunk 5 above)
Comment (sindelio) [reply cg/test/management/subjects.api.spec.js]: It's more intuitive to have an overview of the test in the describe, but I see why it was named like that. I prefer to have a more general overview in the describe, since the it is read separately from the describe most of the time. I can rename it back, what do you think?
